### PR TITLE
curvefs/client: fix fileCacheManager and chunkCacheManager is not released problem

### PR DESCRIPTION
**Pull request by baijiaruo**
…

### What problem does this PR solve?

Issue Number: #742  

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?):

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by baijiaruo** [comment]
image
Stop running mdtest, you can see through metrics that both fileCacheManager and chunkCacheManager have become 0

**Comment by ilixiaocui** [comment]
please add metric of datacache.

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 1]
```diff
@@ -988,7 +984,7 @@ CURVEFS_ERROR FileCacheManager::Flush(bool force, bool toS3) {
                 // so count is 2 if count more than 2,
```
Comment (wuhongsong) [inline curvefs/src/client/s3/client_s3_cache_manager.cpp:984]: comments can be modified

Comment (wuhongsong) [inline curvefs/src/client/s3/client_s3_cache_manager.cpp:984]: Comments can be modified

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 2]
```diff
@@ -195,13 +178,26 @@ CURVEFS_ERROR FsCacheManager::FsSync(bool force) {
...
+                    (iter1->second.use_count() <= 2)) {
                     VLOG(9) << "Release FileCacheManager, inode id: "
                             << iter1->second->GetInodeId();
                     fileCacheManagerMap_.erase(iter1);
                     g_s3MultiManagerMetric->fileManagerNum << -1;
                 }
             }
+        } else if (ret == CURVEFS_ERROR::NOTEXIST) {
+            iter->second->ReleaseCache();
+            WriteLockGuard writeLockGuard(rwLock_);
```
Comment (wuhongsong) [inline curvefs/src/client/s3/client_s3_cache_manager.cpp:190]: may be this function can be more concise

File: curvefs/src/client/s3/client_s3_cache_manager.cpp (hunk 1 above)
Comment (baijiaruo) [reply curvefs/src/client/s3/client_s3_cache_manager.cpp:1000]: fixe

**Comment by baijiaruo** [comment]
> please add metric of datacache.

add datacache metric

**Review by wuhongsong — APPROVED** [review]

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 3]
```diff
@@ -190,18 +203,32 @@ CURVEFS_ERROR FsCacheManager::FsSync(bool force) {
...
                 if ((iter1->second->IsEmpty()) &&
-                    (2 == iter1->second.use_count())) {
+                    (iter1->second.use_count() <= 2)) {
                     VLOG(9) << "Release FileCacheManager, inode id: "
                             << iter1->second->GetInodeId();
                     fileCacheManagerMap_.erase(iter1);
                     g_s3MultiManagerMetric->fileManagerNum << -1;
                 }
             }
+        } else if (ret == CURVEFS_ERROR::NOTEXIST) {
```
Comment (ilixiaocui) [inline curvefs/src/client/s3/client_s3_cache_manager.cpp:219]: Why can it be judged later here?

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 4]
```diff
@@ -966,16 +993,18 @@ CURVEFS_ERROR FileCacheManager::Flush(bool force, bool toS3) {
...
-                // so count is 2 if count more than 2,
+                        << context->chunkCacheManptr.use_count()
+                        << ",inode:" << inode_
+                        << ", index:" << context->chunkCacheManptr->GetIndex();
+                // tmp、chunkCacheMap_ and context->chunkCacheManptr has this
+                // ChunkCacheManagerPtr, so count is 3 if count more than 3,
                 // this mean someone thread has this ChunkCacheManagerPtr
                 if (iter1->second->IsEmpty() &&
-                  (2 == iter1->second.use_count())) {
+                  (iter1->second.use_count() <= 3)) {
```
Comment (ilixiaocui) [inline curvefs/src/client/s3/client_s3_cache_manager.cpp:1003]: If the context also holds chunkCacheManptr, it is released here, will there be no problem?

File: curvefs/src/client/metric/client_metric.h [hunk 5]
```diff
@@ -187,13 +187,21 @@ struct ClientOpMetric {
 struct S3MultiManagerMetric {
     const std::string prefix;
     bvar::Adder<int64_t> fileManagerNum;
-    bvar::Adder<int64_t> chunkCacheNum;
+    bvar::Adder<int64_t> chunkManagerNum;
```
Comment (ilixiaocui) [inline curvefs/src/client/metric/client_metric.h:190]: Let's add the iosize metric at the entrance.Please refer to：https://github.com/opencurve/curve/blob/b107828c0c9c7b118fab53e3f4a4e4ff2702e228/src/client/client_metric.h#L116

File: curvefs/src/client/s3/client_s3_cache_manager.cpp (hunk 3 above)
Comment (baijiaruo) [reply curvefs/src/client/s3/client_s3_cache_manager.cpp:219]: file maybe delete，get indoe return CURVEFS_ERROR::NOTEXIST

File: curvefs/src/client/s3/client_s3_cache_manager.cpp (hunk 4 above)
Comment (baijiaruo) [reply curvefs/src/client/s3/client_s3_cache_manager.cpp:1003]: if other context holds chunkCacheManptr, use count must more than 3

File: curvefs/src/client/metric/client_metric.h (hunk 5 above)
Comment (baijiaruo) [reply curvefs/src/client/metric/client_metric.h:190]: image
image
Run fio read and write metrics data respectively

Comment (ilixiaocui) [reply curvefs/src/client/metric/client_metric.h:190]: 为什么metric name是io_size_latency? 而不是直接为io_size?
Why is the metric name io_size_latency?

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 6]
```diff
@@ -190,18 +203,32 @@ CURVEFS_ERROR FsCacheManager::FsSync(bool force) {
...
                 VLOG(9) << "FileCacheManagerPtr count:"
-                        << iter1->second.use_count();
+                        << iter1->second.use_count()
+                        << ", inodeId:" << iter1->first;
                 // tmp and fileCacheManagerMap_ has this FileCacheManagerPtr, so
                 // count is 2 if count more than 2, this mean someone thread has
                 // this FileCacheManagerPtr
                 if ((iter1->second->IsEmpty()) &&
-                    (2 == iter1->second.use_count())) {
+                    (iter1->second.use_count() <= 2)) {
```
Comment (ilixiaocui) [inline curvefs/src/client/s3/client_s3_cache_manager.cpp:212]: 当前的遗留问题在这里标注一下
//TODO(@xxx): 当前问题 pr链接，还有一个fileCacheManager的count为3，无法删除

File: curvefs/src/client/metric/client_metric.h (hunk 5 above)
Comment (baijiaruo) [reply curvefs/src/client/metric/client_metric.h:190]: This is the latency field that comes with bvar::LatencyRecorder

**Review by ilixiaocui — APPROVED** [review]

**Pull request merged**